### PR TITLE
Create new higher order component for money (accounting-js) related prop conversions.

### PR DESCRIPTION
**Pull request by nerrad**
## Problem this Pull Request solves

A useful tool for dealing with money related value display is `accounting-js`.  We already use this extensively within the current event editor.  While working on the [barcode-scanner shortcode conversion](https://github.com/eventespresso/eea-barcode-scanner/pull/7), I realized that there were various places where values could be displayed as money.  Hence the idea for doing a `withMoney` HOC that allows for exposing component that will automatically have money related props that are passed in as numbers automatically converted to the appropriate money formatted display.

The goal for this initial iteration of the HOC, it's just handling converting designated props to be converted into money value objects. 

The following has been accomplished in this pull:

- Introduction of new `Money` and `Currency` value objects. 
- Building a default `SiteCurrency` currency instance that derives its configuration from data passed thru from the server.
- Introduction of a `withMoney` HOC that allows for converting props to `Money` instances for use in a component requiring Money value objects to be used to represent Money.
- writing tests covering all the above.
- writing documentation covering all the above.

More information about the new functionality can be discovered via the documentation in this pull.

## How has this been tested

* [x] local unit tests
* [x] Testing with implementation on the barcode scanner add-on.

## Checklist

* [ ] I have added a changelog entry for this pull request
* [x] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [x] User input is adequately validated and sanitized
* [x] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [x] My code is tested.
* [x] My code follows the Event Espresso code style.
* [x] My code has proper inline documentation.
* [x] My code accounts for when the site is Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

File: assets/src/vo/currency.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+				'must be a boolean primitive.'
+			);
+		}
+
+		if ( config.decimalPlaces && ! isNumber( config.decimalPlaces ) ) {
+			throw new TypeError(
+				'The decimalPlaces property on the configuration object ' +
+				'must be a number primitive'
+			);
+		}
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/currency.js]: No.  There's actually a country that has currency with 4 decimal places.  However, I think we should leave the range not checked because of the advent of crypto currencies (which can have multiple decimal places.  See [chart here](https://en.wikipedia.org/wiki/ISO_4217).

File: core/domain/services/assets/CoreAssetManager.php [hunk 2]
```diff
@@ -202,12 +204,20 @@ private function loadCoreJs()
...
+            )
+        )->setRequiresTranslation();
+
         $this->addJavascript(
             CoreAssetManager::JS_HANDLE_EE_HOC_COMPONENTS,
             $this->registry->getJsUrl($this->domain->assetNamespace(), 'hocComponents'),
             array(
                 CoreAssetManager::JS_HANDLE_EE_DATA_STORES,
-                CoreAssetManager::JS_HANDLE_EE_HELPERS,
+                CoreAssetManager::JS_HANDLE_EE_VALUE_OBJECTS,
```
[the remark this answers is not in the document]
Comment (nerrad) [reply core/domain/services/assets/CoreAssetManager.php]: correct and yes expected.  It means that we can start using `@eventespresso/model` in our components instead of importing directly (and should lead to smaller bundle size for the components bundle).

File: core/domain/services/assets/CoreAssetManager.php [hunk 3]
```diff
@@ -217,7 +227,7 @@ private function loadCoreJs()
             $this->registry->getJsUrl($this->domain->assetNamespace(), 'components'),
             array(
                 CoreAssetManager::JS_HANDLE_EE_DATA_STORES,
-                CoreAssetManager::JS_HANDLE_EE_HELPERS,
+                CoreAssetManager::JS_HANDLE_EE_VALUE_OBJECTS,
```
[the remark this answers is not in the document]
Comment (nerrad) [reply core/domain/services/assets/CoreAssetManager.php]: Yup (see earlier comment).

File: core/domain/services/assets/CoreAssetManager.php [hunk 4]
```diff
@@ -242,7 +252,7 @@ private function loadCoreJs()
                     'collection_endpoints' => EED_Core_Rest_Api::getCollectionRoutesIndexedByModelName(),
                     'primary_keys' => EED_Core_Rest_Api::getPrimaryKeyNamesIndexedByModelName(),
 					'site_url' => site_url('/'),
-                    'admin_url' => admin_url('/'),
+                    'admin_url' => admin_url('/')
```
[the remark this answers is not in the document]
Comment (nerrad) [reply core/domain/services/assets/CoreAssetManager.php]: unintended, there were iterations of this specific file in this branch while I was working out the new currency config and it probably got accidentally removed when I cut out what I added there.  I'll add back in.

File: docs/AA--Javascript-in-EE/higher-order-components/base-control.md [hunk 5]
```diff
@@ -77,7 +77,7 @@ Note: this package isn't published yet so you'll need to point to it via your bu
 ```js
 module.exports = {
     externals: {
-        '@eventespresso/higher-order-components': 'eejs.higher-order-components',
+        '@eventespresso/higher-order-components': 'eejs.hocComponents',
```
[the remark this answers is not in the document]
Comment (nerrad) [reply docs/AA--Javascript-in-EE/higher-order-components/base-control.md]: Ya its been this way for a while :/ and I just needed to fix the documentation.  We can probably change in a separate pull.

File: docs/AA--Javascript-in-EE/value-objects/money.md [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+This method provides a convenient way to compare the current instance of a money object with another one passed in as an argument.  A money object is considered "equal" when the values on the amount are the same, and a shallow equality check on the properties of the currencies are the same.  Equality does not consider "sameness" in terms of object instance.
+
+```js
+const moneyA = new Money( new Decimal( 1.25 ), new Currency( USDCurrencyConfig ) );
+const moneyB = new Money( new Decimal( 1.25 ), new Currency( USDCurrencyConfig ) );
+const moneyC = new Money( new Decimal( 3.40 ), new Currency( USDCurrencyConfig ) );
+// prints true
+console.log( moneyA.equals( moneyB ) );
+//prints false
+console.log( moneyA.equals( moneyC ) );
```
[the remark this answers is not in the document]
Comment (nerrad) [reply docs/AA--Javascript-in-EE/value-objects/money.md]: The convention for comments is that the comment block comes _before_ the code it refers to (when on separate lines).  I do agree about consistency though so I'll modify the earlier docs that have the comments on the same line.

File: docs/AA--Javascript-in-EE/value-objects/money.md [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+const moneyA = new Money( 1.25, USDCurrency );
+const moneyB = new Money( '1.25', USDCurrency );
+const moneyC = new Money( 100, USDCurrency );
+const moneyD = new Money( .56, new Currency( USDCurrency.toJSON() ) );
+// returns  0
+console.log( moneyA.compare( moneyB ) );
+// returns 1
+console.log( moneyA.compare( moneyC ) );
+// returns -1
+console.log( moneyA.compare( moneyD ) );
```
[the remark this answers is not in the document]
Comment (nerrad) [reply docs/AA--Javascript-in-EE/value-objects/money.md]: Actually both were wrong.  `moneyA` is _less than_ `moneyC` and greater than `moneyD`.  So I just switched the two.  Good catch!

File: assets/src/vo/money.js [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+	}
+};
+
+/**
+ * Asserts if two currencies are shallow equal.
+ * @param {Currency} currencyA
+ * @param {Currency} currencyB
+ * @throws {Exception}
+ */
+const assertEqualCurrency = ( currencyA, currencyB ) => {
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/money.js]: The reason I didn't go with "same" is because I thought that may get confused with it asserting that the two values are the same _instance_.  It doesn't, it asserts if the values represented by the instances are _equal_.  There's a subtle but **important** distinction there.

File: assets/src/vo/money.js [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+	 * @type {number}
+	 */
+	static ROUND_HALF_EVEN = Decimal.ROUND_HALF_EVEN;
+
+	/**
+	 * Rounds towards nearest neighbour. If equidistant, rounds towards
+	 * Infinity.
+	 * @type {number}
+	 */
+	static ROUND_HALF_CEIL = Decimal.ROUND_HALF_CEIL;
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/money.js]: `ROUND_HALF_UP` rounds away from zero when the value is equidistant.  `ROUND_HALF_CEIL` rounds towards **infinity** if equidistant.  That's primarily the description.  I just copied these constants over from the decimal library.  But I agree, there is essentially no difference between the two (and one could argue the description doesn't even make sense (which was also copied from Decimal).  

I've removed both `ROUND_HALF_CEIL` and `ROUND_HALF_FLOOR` as options.  They could always get added back in if/when we discover there _is_ a difference.

File: assets/src/vo/money.js [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+
+	/**
+	 * Rounds towards nearest neighbour. If equidistant, rounds towards
+	 * -Infinity.
+	 * @type {number}
+	 */
+	static ROUND_HALF_FLOOR = Decimal.ROUND_HALF_FLOOR;
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/money.js]: See above.

File: assets/src/vo/money.js [hunk 11]
```diff
@@ -0,0 +1,519 @@
...
+	 * object's Currency.
+	 *
+	 * This does a shallow comparison on the serialized values for the currency
+	 * objects.  That way if the currencies are different instances, but share
+	 * the same internal value, they are considered equal.
+	 *
+	 * @param {Money} other
+	 * @return {boolean} True means the currencies are equal.
+	 */
+	hasEqualCurrency( other ) {
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/money.js]: > using hasEqual sounds like they could actually be different but still somehow have equality in some other sense.

This is exactly why it is this way.  This allows for them to be different _instances_ of an object but still equal in _value_. If someone needs to know whether two instances are identical they can just do a `===` check.  The purpose of this method is to determine equivalency of value (via a shallow compare of the value represented by the currency).  You can view the tests to see the implication of this.

> ie: currencies WXY and XYZ are both trading at a value of 1.25 relative to currency ABC and can therefore be considered equivalent, despite not being the same currency.

This example doesn't really apply here because in that case this method would return `WXY` and `XYZ` as being non-equal.  Again, please look at the tests and it will become clearer how this works.  If after the tests it still isn't clear ping me in chat.

File: assets/src/vo/money.js [hunk 12]
```diff
@@ -0,0 +1,519 @@
...
+
+	/**
+	 * Asserts if the provided values are both Money objects and have Equal
+	 * Currency objects.
+	 *
+	 * @param {Money} thisMoney
+	 * @param {Money} otherMoney
+	 * @throws {TypeError}
+	 */
+	static assertEquivalentWithCurrency = ( thisMoney, otherMoney ) => {
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/money.js]: I explained in an earlier response why I used the term "equal" vs "same" so I won't explain again here.  However, I'm fine with modifying this method name to `assertUsingEqualCurrency`.

File: assets/src/vo/money.js [hunk 13]
```diff
@@ -0,0 +1,519 @@
...
+		assertEqualCurrency( thisMoney.currency, otherMoney.currency );
+	};
+
+	/**
+	 * Asserts if two currencies are shallow equal.
+	 * @param {Currency} currencyA
+	 * @param {Currency} currencyB
+	 * @throws {Exception}
+	 */
+	static assertEqualCurrency = ( currencyA, currencyB ) => {
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/money.js]: As mentioned earlier I disagree (for the reasons already given).

File: assets/src/vo/money.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/money.js]: Also when I talk about value here remember we're talking about the _currency_ NOT the _money_ instance.  So _value_ represents the configuration for that currency.

File: assets/src/vo/currency.js [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+	 * The singular label for the currency (eg. 'Dollar');
+	 * @type {string}
+	 */
+	singularLabel = '';
+
+	/**
+	 * The plural label for the currency (eg. 'Dollars');
+	 * @type {string}
+	 */
+	pluralLabel = '';
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/currency.js]: I'm still not sure whether I'll have the Currency Vo require usage, but I'm going to create a Label value object as a part of the work I'm doing for the hardcoded pretty status labels in another branch (turns out its a good thing to implement there to save having to pick from arrays).

File: assets/src/vo/currency.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/vo/currency.js]: ooo good point about cryptos... hadn't considered that which is not a wise approach these days...

File: assets/src/vo/money.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/vo/money.js]: > please look at the tests and it will become clearer how this works.

I realize exactly how it works and what is actually happening in the method but my issue is that the naming potentially **sounds** like something else is happening. To be clear, I'm somewhat splitting hairs because the terms "equal" and "same" in many cases are interchangeable, however, in this context I don't think they are, and using "equal" potentially conveys something else, whereas using "same" would be unequivocally clear.

Remember that with Value Objects we don't care about instances. We only care about values. So two instances with equal values are not just considered "equal", they are considered to be the **same** thing, which is exactly why I'm suggesting the name change. Anybody doing instance checks with Value Objects is doing it wrong because one of the fundamental cornerstones of the pattern is that individual instances do not matter. An object representing USD is the **same** as any other object representing USD. They don't just have equal values, we literally consider them to be the same thing, so they are completely interchangeable. Two things with "equal" values should not be considered value objects if we are considering them to be separate things, but should instead be considered entities because we are considering their identity. Two Money objects that both use a USD Currency object are considered to be sharing the same Currency, even though it is almost guaranteed that the actual instances are different (common with VOs since new instances are constantly being created). It's a subtle but important difference. Remember, we never ever ever care about instances, so two Money objects with USD Currency are sharing the same Currency object. 

If we converted `hasEqualCurrency()` into a plain sentence/question, it might be something like:

- does this money have an equal currency?
- is the currency for this other amount of money equal to this amount of money?
- are these two money amounts displayed with an equal currency?
- do these two money amounts have an equal currency?

First thing right off the bat... people don't talk like that.
It sounds totally awkward and weird, so why use weird awkward phrasing to name a method?

Secondly, using the term "equal" leads to the question "equal in what way?". 
Using the term "equal" means that we are indeed comparing values. You even stated that outright in your response:

>  The purpose of this method is to determine equivalency of value 

and my beef with that, is that it's not inherently clear whether we doing that for ALL of the values, or only some of them. Using the term "same" makes it clear that ALL of the values are equal and not just a subset.

Thirdly, the above sentences sound like we are potentially talking about more than one currency (ex: do these two money amounts have an equal currency?),  whereas using "same" offers a slight change in meaning and makes it clear that there is only **one** currency in question:

- does this money have the same currency?
- is the currency for this other amount of money the same as this amount of money?
- are these two money amounts displayed with the same currency?
- do these two money amounts have the same currency?

Here's another way to approach this...
let's put both terms into one sentence and separate them with an "OR" to determine which one better conveys the concept of to things being identical:

### is this an equal currency or is this the same currency?

one of those terms unequivocally conveys the idea that we are talking about something that is identical in every way (remember we don't care about instances).

bottom line: `hasEqualCurrency()` is just not as unequivocally clear as `hasSameCurrency()`

File: assets/src/vo/money.js [hunk 15]
```diff
@@ -237,7 +223,7 @@ export default class Money {
 	 * @return {Money} Returns a new instance of Money.
 	 */
 	add( other ) {
-		Money.assertEquivalentWithCurrency( this, other );
+		Money.assertUsingEqualCurrency( this, other );
```
Comment (tn3rb) [inline assets/src/vo/money.js]: better... but still think `assertUsingSameCurrency()` is even **more** clear that there are zero differences

File: docs/AA--Javascript-in-EE/value-objects/money.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (tn3rb) [reply docs/AA--Javascript-in-EE/value-objects/money.md]: OH DOH... ya `moneyC` is 100... 
I was thinking in subunits for that one for some dumb reason... 
guess I was bamboozled by the example claiming that 1.25 > 100 and just drank the kool-aid

File: assets/src/vo/currency.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/vo/currency.js]: right... I've thought many times about using a Label VO in our PHP code, but it would have to replace existing code which causes all kinds of issues.
Since this stuff is all starting out from scratch I felt I should mention it before it is too late to change. The main benefits of using a VO for this is it standardizes our usage of labels for things, and can potentially make some process easier.

That said, not using a VO for labels doesn't complicate things much so not going to push for this.

File: assets/src/vo/money.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/src/vo/money.js]: It's clear from the thread so far that we both think the choice of term matters. Essentially if I summarize the conflicting positions here:

### My argument

Using the term "same" (within code) is equivalent to `===` or strict equality because it refers to something that is identical (not only the same value, but also the same type or instance).  Using the term "equals" (in code) is equivalent to `==` or loose equality because it refers to something that has the same value (but may not be the same type or instance).

**Since its true we don't care about whether the instances are the same** in value objects and all we care about is whether the value is equal, I chose to use the term equals.  I think for the purposes of the methods described by this term it is clear what it is doing.

### Your Argument

`hasEqualCurrency()` is just not as unequivocally clear as `hasSameCurrency()` because you believe that `hasSameCurrency()` conveys the idea that we are talking about something that is _identical in every way_ (because we don't care about instances).  

Essentially the disagreement (as I see it) boils down to you saying that the meaning of "same" (or identical) in the context of vo's is different than in any other code context because things that might disqualify something from being the same outside a vo are not applicable to inside a vo.  Whereas my point is that I don't want to change the meaning of a term to fit the context, but rather use the meaning of a term that is _still useful_ in this context.  

Put another way.  When comparing value objects, it doesn't matter if they are completely identical (equal in instance AND value), what matters is they are _equal_ in value only.  So I chose the term "equals".  

As for your plain english examples, while in some cases trying to sound out how we'd talk about code in plain English can help clarify things, I think it also can be misleading because English is not a very precise language.  In every day conversation, currency is more of an abstract concept.  People use currency to describe the kind of money they have, "I get paid in US Dollars, not Canadian dollars".  So in _everyday conversation_ the concept of "same" or "equal" applies to the dollar as a unit.  I'm not going to say "Does this dollar have an equal (or same) currency as that dollar?"  No, we're more likely to say "The Canadian dollar is not equal in value (or the same in value) as the US dollar".  In everyday conversation there's essentially no difference between usage of the terms "same" and "equals".

However when we get to writing code, the choice of terms has more meaning, and that's why I think it matters what we choose to describe the behaviour of a method.   I _know_ instances don't matter in value objects, so I'm not going to use the term **same** for a method that compares the _values_ of two instances.   Instead I chose what I feel is the _still useful_ and more accurate - the term **equals**. Functionally that's all we care about.

[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/vo/money.js]: > Using the term "same" (within code) is equivalent to === or strict equality because it refers to something that is identical (not only the same value, but also the same type or instance). 

exactly... and since we are talking about VOs you absolutely have to drop the concept of instances because they don't exist with VOs. If all of the values are the same, then they are identical. It's true that you can create multiple instances of a VO **in code** (which is in fact a really common occurrence with VOs since that's how you maintain immutability), but you should never ever ever consider those instances in any way whatsoever. You basically have to ignore their existence. Why? Because context totally matters when you are attempting to model something from the real world. Just because something may be true with regards to the code you write, does not make it true in the real world. There is only one USD currency. Trying to conceptualize multiple USD currencies makes as much sense as saying something is 1 km north of the North pole. You can't have 1 instance of USD currency in one hand and another instance of USD currency in the other, at least not without bringing in the concept of money itself. It doesn't make sense in the real world. You also can't add the USD Currency to itself (USD + USD = how many USD !?!?), but we could do it in code right? We could create thousands of instances of USD Currency in code, but in the real world, there is only ever going to be **one** USD Currency, and that's what we are trying to model. USD Currency is USD Currency is USD Currency. They are all identical to any other USD Currency, ie: 

```
USD Currency === USD Currency
```

two amounts of USD money share the same currency. We don't say their currencies are equal because that implies there is more than one currency in play. ie: "the currency for this $10 USD is equal to the currency for this other $10 USD". Thinking like that doesn't make sense.

> Using the term "equals" (in code) is equivalent to == or loose equality because it refers to something that has the same value 

but loose equality is not necessarily the same value depending on the **context** of what you are referring to. Look at PHP's "truthy" values:

```php
1 == "1" == true // all loosely equal
0 == "0" == false == null // all loosely equal
```

sure those values are certainly interchangeable in a lot of places (if not most)... but NOT everywhere... the fact is... context DOES matter.

> (but may not be the same type or instance). 

**This is exactly my point!!!**

The term "equals" can potential mean that two things (and this is right from your own wording): **may not be the same**
And again, because we are talking about a currency VO, we have to completely ignore instances (because, you know, real world), so conveying that something may potentially not be equal (while ignoring instances) means that one of the values **may not be the same**.

This is why I'm being so insistent about using "same" vs "equal" because with regards to currency, we don't want to convey that things are "loosely equivalent, but may not be the same" we want to convey that they are identical.

> my point is that I don't want to change the meaning of a term to fit the context

but again, context totally matters.

Here's something that may (or may not) blow your mind (or at least make you go "huh")... for our purposes, money can be conceptualized as a value object, because again, we don't care about instances, and only care about amounts and what currency those amounts are in. Nearly all stores and businesses fall into this category. If however, we were writing software for a central bank (or the Federal Reserve in the US) then we would want to conceptualize money as an Entity and suddenly the concept of identity would become an important property. At this point though we would probably be modeling "Notes", or "Bills", and would use the serial numbers printed on physical money for the individual identifiers. This is one of the main conceptual tenants of Domain Driven Design, which is that your data schema is modeled entirely around the domain, which is the real world space being emulated, and the meanings of all terms depend on the context they are used within.

So in one domain, Money can be modeled as a value object with no identity and the concept of equality depends solely on having matching amounts and currency, and in another domain, Money needs to be modeled as an entity where each item has a separate identity, and the concept of identical equality likely infers the presence of a federal crime (because no two bills should have the exact same serial number). Context totally matters and the concept of equality changes depending on the context.

> As for your plain english examples, while in some cases trying to sound out how we'd talk about code in plain English can help clarify things, I think it also can be misleading because English is not a very precise language. In every day conversation, currency is more of an abstract concept. People use currency to describe the kind of money they have, "I get paid in US Dollars, not Canadian dollars". So in everyday conversation the concept of "same" or "equal" applies to the dollar as a unit. I'm not going to say "Does this dollar have an equal (or same) currency as that dollar?" No, we're more likely to say "The Canadian dollar is not equal in value (or the same in value) as the US dollar". In everyday conversation there's essentially no difference between usage of the terms "same" and "equals".

mmmmm... ya not sure what your point is since you kinda switched over to talking about money and/or exchange rates and not currency on its own. 
If we stick to talking strictly about currency as a separate concept that describes money, then we CAN easily imagine an everyday conversation that uses it within that context:

> store clerk: Should I put this $10 bill in with all of the other $10 bills?
> manager: If it's the same currency, then ya of course. If it's not the same currency, then put it under the money tray.

Now say that last sentence but use "equal" instead:

> manager: If it's an equal currency, then ya of course. If it's not an equal currency, then put it under the money tray.

ya... nobody is going to talk like that, although to be fair they would probably same something like:

> manager: If it's (reference to country) $10, then ya of course. If it's not (reference to country), then put it under the money tray.

So with regards to everyday conversation and/or plain language usage nobody is going to favour the term "equals" over "same" when talking about a currency. With regards to the technical aspect of things, the only way that "equals" makes more sense is if we introduce the concept of instances because then the inference of something being identical that comes with the term "same" wouldn't be applicable to the Currency VO. But that is totally doing it wrong because you should not consider instances when dealing with value objects because they do not have identity. The argument for this is basically saying:

> the term equals is more accurate... because if we were referring to entities with separate identities and the individual instances were important... then it would make sense!

but we are not referring to entities with separate identities, we're referring to value objects and the individual instances should be disregarded!

This argument is equivalent to:

> Why are you driving on the wrong side of the road?
> If we were in a country where everyone drove on this side of the road then it would make sense to drive on this side
> ...uhhhh... but we're not in one of those countries...

Same issue here. If we were talking about entities, then instances and identities, and the concepts of equality and sameness would take on different meanings because the context would be different. That's not the case here. Similarly, if we were attempting to model an entirely different domain, then we would be dealing with different contexts and may have to approach familiar concepts from an entirely different angle.

Anyways, I'm sorry for being a pain and being so stubborn about this but I think naming is important and want to ensure that things have the most intuitive names they can have. Once this gets out there it will be hard to change so let's get it perfect now. I think I'll post something in chat to get some feedback from the rest of the team with regards to what they feel would be a more intuitive naming schema.

[the remark this answers is not in the document]
Comment (sethshoultes) [reply assets/src/vo/money.js]: > I think I'll post something in chat to get some feedback from the rest of the team with regards to what they feel would be a more intuitive naming schema.

A Grinder post might be better for discussion.

**Comment by tn3rb** [comment]
Added a poll in chat regarding which method name was more clear.
You can see the results here: https://eventespresso.slack.com/archives/C02SY781M/p1532969445000488
Plz note that I was very careful not to word the question in a way that would favour one option over the other.

The results were:

`hasEqualCurrency()` 1 vote (yours)

`hasSameCurrency()` 6 votes (mine + 5 others)

So I'm sorry but everybody that responded appeared to agree that at least with regards to currency, "same" is a more clear descriptor to use.

File: docs/AA--Javascript-in-EE/value-objects/money.md [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+
+#### `Money.assertDecimal`
+```js
+Money.assertDecimal( maybeDecimal );
+```
+This method throws a `TypeError` if the provided argument is not an instance of `Decimal`.
+
+#### `Money.assertSameCurrency`
+```js
+Money.assertEqualCurrency( currencyA, currencyB );
```
Comment (tn3rb) [inline docs/AA--Javascript-in-EE/value-objects/money.md]: needs updating

File: docs/AA--Javascript-in-EE/value-objects/money.md [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+```js
+const moneyA = new Money( new Decimal( 1.25 ), new Currency( USDCurrencyConfig ) );
+const moneyB = new Money( new Decimal( 2.35 ), new Currency( USDCurrencyConfig ) );
+const moneyC = new Money( new Decimal( 400 ), new Currency( CDNCurrencyConfig ) );
+
+// prints true
+console.log( moneyA.hasEqualCurrency( moneyB ) );
+
+//prints false
+console.log( moneyA.hasEqualCurrency( moneyC ) );
```
Comment (tn3rb) [inline docs/AA--Javascript-in-EE/value-objects/money.md]: needs updating

File: docs/AA--Javascript-in-EE/value-objects/money.md [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+#### `hasSameCurrency()`
+This method provides a convenient way to compare the currencies on the current Money instance with the currency on another Money instance provided as an argument.  Currencies are considered equal via a shallow equality check on the properties of the two Currency instances in each Money instance.
+
+```js
+const moneyA = new Money( new Decimal( 1.25 ), new Currency( USDCurrencyConfig ) );
+const moneyB = new Money( new Decimal( 2.35 ), new Currency( USDCurrencyConfig ) );
+const moneyC = new Money( new Decimal( 400 ), new Currency( CDNCurrencyConfig ) );
+
+// prints true
+console.log( moneyA.hasEqualCurrency( moneyB ) );
```
Comment (tn3rb) [inline docs/AA--Javascript-in-EE/value-objects/money.md]: needs updating

Comment (nerrad) [reply docs/AA--Javascript-in-EE/value-objects/money.md]: 👍
